mne/decoding/time_gen.py

```diff
@@ -387,40 +388,88 @@ def _predict_time_loop(X, estimators, cv, slices, predict_mode):
     """
     n_epochs = len(X)
     # Loop across testing slices
-    y_pred = [list() for _ in range(len(slices))]
 
-    # XXX EHN: This loop should be parallelized in a similar way to fit()
+    if predict_mode == 'cross-validation':
+        # Check that training cv and predicting cv match
+        if (len(estimators) != len(cv)) or (cv.n != n_epochs):
+            raise ValueError(
+                'When `predict_mode = "cross-validation"`, the training '
+                'and predicting cv schemes must be identical.')
+        all_test = np.concatenate(list(zip(*cv))[-1])
+        test_slices = []
+        start = 0
+        for _, test in cv:
+            l = len(test)
+            stop = start + l
+            test_slices.append(slice(start, stop, 1))
+            start += l
+
+    if all(len(s) == 1 for s in slices):
+        slices = [slice(s[0], s[0] + 1, 1) for s in slices]
+
+    y_pred = list()
+
+    # check whether the GAT is based on window length = 1-time-sample
+    # Note that we have to check i) that slices' step = 1 and ii) that slices
+    # are consecutive.
+    is_single_time_sample = True
+    expected_start = np.arange(len(slices))
+    expected_stop = expected_start + 1
+    if any(not isinstance(sl, slice) for sl in slices):
+        is_single_time_sample = False
+    elif not np.array_equal([sl.start for sl in slices], expected_start):
+        is_single_time_sample = False
+    elif not np.array_equal([sl.stop for sl in slices], expected_stop):
+        is_single_time_sample = False
+    elif slices[-1].stop != X.shape[-1]:
+        is_single_time_sample = False
```

Comment:
Do you guys find that more readable? I prefer the boring version that we have there.

> On 06 Dec 2015, at 15:28, jona-sassenhagen <notifications@github.com> wrote:
> 
> In mne/decoding/time_gen.py:
> 
> > +    y_pred = list()
> > +
> > +    # check whether the GAT is based on window length = 1-time-sample
> > +    # Note that we have to check i) that slices' step = 1 and ii) that slices
> > +    # are consecutive.
> > +    is_single_time_sample = True
> > +    expected_start = np.arange(len(slices))
> > +    expected_stop = expected_start + 1
> > +    if any(not isinstance(sl, slice) for sl in slices):
> > +        is_single_time_sample = False
> > +    elif not np.array_equal([sl.start for sl in slices], expected_start):
> > +        is_single_time_sample = False
> > +    elif not np.array_equal([sl.stop for sl in slices], expected_stop):
> > +        is_single_time_sample = False
> > +    elif slices[-1].stop != X.shape[-1]:
> > +        is_single_time_sample = False
> is_single_time_sample = (
>     all(isinstance(sl, slice) for sl in slices)
>     np.array_equal([sl.start for sl in slices], expected_start) or
>     np.array_equal([sl.stop for sl in slices], expected_stop) or
>     slices[-1].stop == X.shape[-1]) 
> ?
> 
> —
> Reply to this email directly or view it on GitHub.
>